internal/providers/gitlab/registration.go

```diff
@@ -0,0 +1,172 @@
+)
+
+// RegisterEntity implements the Provider interface
+func (c *gitlabClient) RegisterEntity(
+	ctx context.Context, entType minderv1.Entity, props *properties.Properties,
+) (*properties.Properties, error) {
+	if !c.SupportsEntity(entType) {
+		return nil, errors.New("unsupported entity type")
+	}
+
+	upstreamID := props.GetProperty(properties.PropertyUpstreamID).GetString()
+	if upstreamID == "" {
+		return nil, errors.New("missing upstream ID")
+	}
+
+	whprops, err := c.createWebhook(ctx, upstreamID)
+	if err != nil {
+		// There is already enough context in the error message
+		return nil, err
+	}
+
+	return props.Merge(whprops), nil
+}
+
+// DeregisterEntity implements the Provider interface
+func (c *gitlabClient) DeregisterEntity(
+	ctx context.Context, entType minderv1.Entity, props *properties.Properties,
+) error {
+	if !c.SupportsEntity(entType) {
+		return errors.New("unsupported entity type")
+	}
+
+	upstreamID := props.GetProperty(properties.PropertyUpstreamID).GetString()
+	if upstreamID == "" {
+		return errors.New("missing upstream ID")
+	}
+
+	hookID := props.GetProperty(RepoPropertyHookID).GetString()
+	if hookID == "" {
+		return errors.New("missing hook ID")
+	}
+
+	if err := c.deleteWebhook(ctx, upstreamID, hookID); err != nil {
+		// There is already enough context in the error message
+		return err
+	}
+
+	return nil
+}
+
+func (c *gitlabClient) createWebhook(ctx context.Context, upstreamID string) (*properties.Properties, error) {
+	createHookPath, err := url.JoinPath("projects", upstreamID, "hooks")
+	if err != nil {
+		return nil, fmt.Errorf("failed to join URL path for hooks: %w", err)
+	}
+
+	trve := ptr.Ptr(true)
+	hreq := &gitlab.AddHookOptions{
+		URL: &c.webhookURL,
+		// TODO: Add secret
+		PushEvents:             trve,
+		TagPushEvents:          trve,
+		MergeRequestsEvents:    trve,
+		RepositoryUpdateEvents: trve,
+		// TODO: Enable SSL verification
+	}
+
+	hook, err := c.doCreateWebhook(ctx, createHookPath, hreq)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create webhook: %w", err)
+	}
+
+	outProps, err := properties.NewProperties(map[string]interface{}{
+		// we store as string to avoid any type issues. Note that we
+		// need to retrieve it as a string as well.
+		RepoPropertyHookID:  fmt.Sprintf("%d", hook.ID),
+		RepoPropertyHookURL: hook.URL,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create properties: %w", err)
+	}
+
+	return outProps, nil
+}
+
+func (c *gitlabClient) doCreateWebhook(
+	ctx context.Context, createHookPath string, hreq *gitlab.AddHookOptions,
+) (*gitlab.Hook, error) {
+	req, err := c.NewRequest(http.MethodPost, createHookPath, hreq)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %w", err)
+	}
+
+	resp, err := c.Do(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get projects: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
```

Comment:
if you dereference resp without checking it and before checking for err it might be `nil`